File: levels/level11/task.py
# Aufgabe: Implementiere den Quick Sort Algorithmus
# Dein Code beginnt hier

import time
import random
from memory_profiler import memory_usage

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

def measure_memory_and_time(func, arr):
    start_time = time.time()
    mem_usage = memory_usage((func, [arr.copy()]), interval=0.01)
    end_time = time.time()
    return (max(mem_usage) - min(mem_usage)) * 1024, (end_time - start_time) * 1000

# Teste den Quick Sort Algorithmus und analysiere Zeit- und Raumkomplexität
if __name__ == "__main__":
    arr = [3, 6, 8, 10, 1, 2, 1]
    large_arr = [random.randint(1, 1000) for _ in range(3000)]  # Listengröße

    # Teste den Algorithmus
    print("Sortierte Liste:", quick_sort(arr))

    # Analysiere Zeit- und Raumkomplexität mit großer Liste
    quick_sort_mem, quick_sort_time = measure_memory_and_time(quick_sort, large_arr)
    print("\nTest mit großer Liste:")
    print("Quick Sort Zeit:", format(quick_sort_time, ".3f"), "ms")
    print("Quick Sort Speicher:", format(quick_sort_mem, ".3f"), "KB")

    # Zusammenfassung der Ergebnisse
    print("\nZusammenfassung der Ergebnisse:")
    print(f"Quick Sort Zeit (Python): {format(quick_sort_time, '.3f')} ms")
    print(f"Quick Sort Speicher (Python): {format(quick_sort_mem, '.3f')} KB")

File: levels/level11/task.cpp
// Aufgabe: Implementiere den Quick Sort Algorithmus
// Dein Code beginnt hier

#include <iostream>
#include <vector>
#include <chrono>
#include <iomanip>
#include <algorithm>
#include <stdlib.h>

void quick_sort(std::vector<int>& arr, int left, int right) {
    int i = left, j = right;
    int pivot = arr[(left + right) / 2];
    while (i <= j) {
        while (arr[i] < pivot)
            i++;
        while (arr[j] > pivot)
            j--;
        if (i <= j) {
            std::swap(arr[i], arr[j]);
            i++;
            j--;
        }
    }
    if (left < j)
        quick_sort(arr, left, j);
    if (i < right)
        quick_sort(arr, i, right);
}

void print_memory_usage() {
    // This function is designed to be used with valgrind for more precise memory usage measurements.
    std::cout << "Memory usage can be measured with tools like valgrind." << std::endl;
}

int main() {
    std::vector<int> arr = {3, 6, 8, 10, 1, 2, 1};
    quick_sort(arr, 0, arr.size() - 1);
    std::cout << "Sortierte Liste: ";
    for (int x : arr) {
        std::cout << x << " ";
    }
    std::cout << std::endl;

    std::vector<int> large_arr(3000);  // Listengröße
    std::generate(large_arr.begin(), large_arr.end(), []() { return rand() % 1000; });

    // Test mit großer Liste
    std::cout << "\nTest mit großer Liste:" << std::endl;

    auto start_time = std::chrono::high_resolution_clock::now();
    quick_sort(large_arr, 0, large_arr.size() - 1);
    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> quick_sort_time = end_time - start_time;
    std::cout << "Quick Sort Zeit: " << std::fixed << std::setprecision(3) << quick_sort_time.count() << " ms" << std::endl;

    print_memory_usage(); // Platzhalter, sollte mit valgrind gemessen werden

    // Zusammenfassung der Ergebnisse
    std::cout << "\nZusammenfassung der Ergebnisse:" << std::endl;
    std::cout << "Quick Sort Zeit (C++): " << std::fixed << std::setprecision(3) << quick_sort_time.count() << " ms" << std::endl;

    return 0;
}
